server/src/main/java/org/elasticsearch/index/translog/TruncateTranslogAction.java

```diff
@@ -0,0 +1,236 @@
+        }
+
+        List<IndexCommit> commits;
+        try {
+            terminal.println("Reading translog UUID information from Lucene commit from shard at [" + indexPath + "]");
+            commits = DirectoryReader.listCommits(indexDirectory);
+        } catch (IndexNotFoundException infe) {
+            throw new ElasticsearchException("unable to find a valid shard at [" + indexPath + "]", infe);
+        }
+
+        // Retrieve the generation and UUID from the existing data
+        commitData = commits.get(commits.size() - 1).getUserData();
+        final String translogGeneration = commitData.get(Translog.TRANSLOG_GENERATION_KEY);
+        final String translogUUID = commitData.get(Translog.TRANSLOG_UUID_KEY);
+        if (translogGeneration == null || translogUUID == null) {
+            throw new ElasticsearchException("shard must have a valid translog generation and UUID but got: [{}] and: [{}]",
+                translogGeneration, translogUUID);
+        }
+
+        final long globalCheckpoint = commitData.containsKey(SequenceNumbers.MAX_SEQ_NO)
+            ? Long.parseLong(commitData.get(SequenceNumbers.MAX_SEQ_NO))
+            : SequenceNumbers.UNASSIGNED_SEQ_NO;
+
+        terminal.println("Translog Generation: " + translogGeneration);
+        terminal.println("Translog UUID      : " + translogUUID);
+        terminal.println("History UUID       : " + historyUUID);
+
+        Path tempEmptyCheckpoint = translogPath.resolve("temp-" + Translog.CHECKPOINT_FILE_NAME);
+        Path realEmptyCheckpoint = translogPath.resolve(Translog.CHECKPOINT_FILE_NAME);
+        Path tempEmptyTranslog = translogPath.resolve("temp-" + Translog.TRANSLOG_FILE_PREFIX +
+            translogGeneration + Translog.TRANSLOG_FILE_SUFFIX);
+        Path realEmptyTranslog = translogPath.resolve(Translog.TRANSLOG_FILE_PREFIX +
+            translogGeneration + Translog.TRANSLOG_FILE_SUFFIX);
+
+        // Write empty checkpoint and translog to empty files
+        long gen = Long.parseLong(translogGeneration);
+        int translogLen = writeEmptyTranslog(tempEmptyTranslog, translogUUID);
+        writeEmptyCheckpoint(tempEmptyCheckpoint, translogLen, gen, globalCheckpoint);
+
+        terminal.println("Removing existing translog files");
+        IOUtils.rm(translogFiles.toArray(new Path[]{}));
+
+        terminal.println("Creating new empty checkpoint at [" + realEmptyCheckpoint + "]");
+        Files.move(tempEmptyCheckpoint, realEmptyCheckpoint, StandardCopyOption.ATOMIC_MOVE);
+        terminal.println("Creating new empty translog at [" + realEmptyTranslog + "]");
+        Files.move(tempEmptyTranslog, realEmptyTranslog, StandardCopyOption.ATOMIC_MOVE);
+
+        // Fsync the translog directory after rename
+        IOUtils.fsync(translogPath, true);
+    }
+
+    private boolean isTranslogClean(ShardPath shardPath, String translogUUID) throws IOException {
+        // TODO: perform clean check of translog instead of corrupted marker file
```

Comment:
This `TODO` seems to be done?

Reply:
yes and no: check for corruption marker for translog has to be there.

fine to remove `TODO` to avoid misunderstanding